### ISSUE TITLE
Build from source installs unit files into incorrect directory

### ISSUE THREAD

**Issue by enkore** [libratbag/libratbag#1019]
```
ninja -C builddir
sudo ninja -C builddir install
```
This installs the unit file to `/usr/local/lib/systemd` ("Installing /home/.../libratbag/builddir/ratbagd.service to /usr/local/lib/systemd"), which isn't actually part of systemd's search path for units (as of sysemd 245 / 245.4-2-arch). The unit should be installed to the `/usr/local/lib/systemd/system` directory instead. Moving it manually there makes it work.

**Comment by enkore** [comment]
Similarly the DBus files don't seem to be picked up, although they seem to be in the correct directory. Even restarting dbus and the machine doesn't make it work and causes sd_bus_request_name() to fail with EACCES (found through printf-debugging, since it by default only reports that it failed). Installing the distro's package but not enabling the service has DBus immediately pick up the necessary XML to make it work, though.

**Comment by FFY00** [comment]
We have custom handling for this in meson which is screwing this up, ugh. pkg-config works:

```
$ pkgconf --define-variable=rootprefix=/usr/local --variable=systemdsystemunitdir systemd
/usr/local/lib/systemd/system
```

@whot did you mean the opposite here? If it doesn't match you warn and change to the intended directory, otherwise you use the default?

```meson
if get_option('prefix') == '/usr' and intended_unitdir != default_unitdir
	message('''
	systemd unitdir libdir mismatch detected, changing unitdir to
		@0@
	or specify with
		mesonconf -Dsystemd-unit-dir=<path>
	See https://github.com/libratbag/libratbag/issues/188
	'''.format(default_unitdir))
	unitdir = default_unitdir
else
	unitdir = intended_unitdir
endif
```

**Comment by FFY00** [comment]
I would also like to take this opportunity to propose removing this logic altogether, if fedora is doing something special, they can fix system's pkgconfig files. I am fine leaving a warning, but having any kind of proactive path handling just because fedora is special is bad IMO. This is the exact problem meson is supposed to solve, right now it just feels like we are writing CMake and hoping to get it right.

**Comment by whot** [comment]
/me tries to remember

Basic problem is that `pkgconfig` always gives you the system directories. Where you install into a custom prefix like `/opt/libratbag` (so you can have multiple versions simultaneously),  you don't want the unit file to end up in `/usr` - that kinda defeats the whole point of a custom prefix. So we need to guess where we should install the file instead given the prefix. That guess is `<prefix>/lib/...` but that doesn't work on those distros that use `lib64` which are Fedora/RHEL and I think the SuSEs.

The specific bug here may be just a missing directory name though:
```
intended_unitdir = join_paths(get_option('prefix'), get_option('libdir'), 'systemd')
```
That line is missing the extra `system` directory to match the `systemdunitdir`.

I think there are multiple issues conflated in this code, see #188.
- systemd file installed in wrong directory when prefix == `/usr`. That would be a bug in the pkgconfig file
- systemd file installed in wrong directory when prefix == `/custom/path`. We want the systemd file to be in `/custom/path/lib/systemd/system` then so the tree is a mirror of the normal tree.

Even with the first bug fixed, this would still install into `/custom/path/lib64/systemd/system` which isn't what we want either. I agree that for cleanliness it's better to force use of  `-Dsystemd-unit-dir` if need be.

**Comment by FFY00** [comment]
> Basic problem is that `pkgconfig` always gives you the system directories.

Not quite, that is true if you just set `prefix`. You need to also set `rootprefix`, as I showed above. We need to make sure `rootprefix` is one of the prefixes systemd will actually search for unit files.

```meson
dep_systemd.get_pkgconfig_variable('systemdsystemunitdir', define_variable: ['rootprefix', get_option('prefix')])
```

I think we should replace the whole logic with that. But we should make sure `prefix` is one of the valid paths (`/usr`, `/usr/local`, `$XDG_DATA_HOME`, `$HOME/.local`). We can also add an option for the users to change `rootprefix` if they wish.

What do you think?

**Comment by FFY00** [comment]
FYI, the pkg-config definition:

```
prefix=/usr
rootprefix=/usr
sysconfdir=/etc
systemdutildir=${rootprefix}/lib/systemd
systemdsystemunitdir=${rootprefix}/lib/systemd/system
...
```

**Comment by whot** [comment]
> You need to also set rootprefix, as I showed above.

Yeah, I tried this before and hadn't known about it, thanks. It works and we can do that. I'd go for the simple solutions of either forcing everyone to set `-Dsystemd-unit-dir` if they want a non-default prefix path, or (my preference) set the `rootprefix` as you suggest above and forcing everyone to use `-Dsystemd-unit-dir` if that doesn't match. Both of these are fairly predictable "you get a reasonable standard behaviour" approaches.

Anything else is getting too involved with custom use-cases you don't want to fix one-by-one. You cannot know where systemd looks for the paths because someone who installs into `/opt/libratbag` may also have a custom systemd in `/opt/systemd` and then the lookup paths are wrong anyway :) And `rootprefix` is not our variable, so having an option to change that means we now have an option on top of systemd's API which just opens cans of worms we don't need to deal with. I'd say stick with one of the simple but predictable options.

**Comment by cengiz-io** [comment]
Sorry for resurrecting this issue but it still seems to be present in current HEAD revision. 

What's the suggested workaround for this?

**Comment by staticssleever668** [comment]
@cengiz-io, hi!
Have you tried specifying `-Dsystemd-unit-dir=/path` when setting up meson project?
E.g. `meson setup build --prefix /usr -Dsystemd-unit-dir=/usr/lib/systemd/systemd`.

**Comment by cengiz-io** [comment]
@staticssleever668 
`meson setup build --prefix /usr/local -Dsystemd-unit-dir=/usr/lib/systemd/system` works fine.
(notice the change in prefix and `system` instead of `systemd`)
